File: treessence/src/main/java/fr/bipi/tressence/sentry/SentryEventTree.java
package fr.bipi.tressence.sentry;

import android.util.Log;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import fr.bipi.tressence.base.PriorityTree;
import io.sentry.Sentry;
import io.sentry.event.Event;
import io.sentry.event.EventBuilder;

/**
 * Logger that will log to sentry server
 */
public class SentryEventTree extends PriorityTree {

    private static final String KEY_TAG = "tag";
    private static final String DEFAULT_PLATFORM = "Android";

    /**
     * Threads from where sentry events will be sent.
     * <p>
     * Threads are needed because if connectivity is bad, then {@link Sentry#capture(Event)} can take a long time to
     * complete
     */
    private final ExecutorService executorService = Executors.newFixedThreadPool(1);

    /**
     * @param priority priority from witch log will be logged
     */
    public SentryEventTree(int priority) {
        super(priority);
    }

    @Override
    protected void log(final int priority, final String tag, @NotNull final String message, final Throwable t) {
        if (skipLog(priority, tag, message, t)) {
            return;
        }

        executorService.submit(new Runnable() {
            @Override
            public void run() {

                EventBuilder eb = new EventBuilder();
                eb.withTag(KEY_TAG, tag)
                    .withLevel(fromAndroidLogPriorityToSentryLevel(priority))
                    .withMessage(message);

                Sentry.capture(eb);

                if (t != null) {
                    Sentry.capture(t);
                }
            }
        });
    }

    private Event.Level fromAndroidLogPriorityToSentryLevel(int priority) {
        switch (priority) {
            case Log.INFO:
                return Event.Level.INFO;
            case Log.WARN:
                return Event.Level.WARNING;
            case Log.ERROR:
                return Event.Level.ERROR;
            case Log.ASSERT:
                return Event.Level.FATAL;
            case Log.DEBUG:
            case Log.VERBOSE:
            default:
                return Event.Level.DEBUG;
        }
    }
}

File: treessence/src/main/java/fr/bipi/treessence/sentry/SentryEventTree.kt
package fr.bipi.treessence.sentry

import android.util.Log
import fr.bipi.treessence.base.PriorityTree
import fr.bipi.treessence.common.filters.Filter
import fr.bipi.treessence.common.filters.NoFilter
import io.sentry.Sentry
import io.sentry.event.Event
import io.sentry.event.EventBuilder
import java.util.concurrent.Executors

/**
 * Logger that will log to sentry server
 *
 * @param priority priority from witch log will be logged
 */
class SentryEventTree @JvmOverloads constructor(
    priority: Int,
    filter: Filter = NoFilter.INSTANCE
) : PriorityTree(priority, filter) {
    /**
     * Threads from where sentry events will be sent.
     *
     *
     * Threads are needed because if connectivity is bad, then [Sentry.capture] can take a long time to
     * complete
     */
    private val executorService = Executors.newFixedThreadPool(1)
    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
        if (skipLog(priority, tag, message, t)) {
            return
        }
        executorService.submit {
            val eb = EventBuilder()
            eb.withTag(KEY_TAG, tag)
                .withLevel(fromAndroidLogPriorityToSentryLevel(priority))
                .withMessage(message)
            Sentry.capture(eb)
            if (t != null) {
                Sentry.capture(t)
            }
        }
    }

    private fun fromAndroidLogPriorityToSentryLevel(priority: Int): Event.Level {
        return when (priority) {
            Log.INFO -> Event.Level.INFO
            Log.WARN -> Event.Level.WARNING
            Log.ERROR -> Event.Level.ERROR
            Log.ASSERT -> Event.Level.FATAL
            Log.DEBUG, Log.VERBOSE -> Event.Level.DEBUG
            else -> Event.Level.DEBUG
        }
    }

    companion object {
        private const val KEY_TAG = "tag"
    }
}
